File: lexer/src/main/java/net/jr/lexer/basicterminals/CCharacter.java
package net.jr.lexer.basicterminals;

import net.jr.lexer.Lexemes;
import net.jr.lexer.automaton.DefaultAutomaton;
import net.jr.lexer.impl.CharConstraint;
import net.jr.lexer.impl.TerminalImpl;

import java.io.IOException;

public class CCharacter extends TerminalImpl {

    public CCharacter() {
        setName("cCharacter");
        DefaultAutomaton.Builder builder = DefaultAutomaton.Builder.forTokenType(this);
        DefaultAutomaton.Builder.BuilderState init = builder.initialState();
        DefaultAutomaton.Builder.BuilderState gotFirstQuote = builder.newNonFinalState();
        DefaultAutomaton.Builder.BuilderState escaped = builder.newNonFinalState();
        DefaultAutomaton.Builder.BuilderState octalEscape = builder.newNonFinalState();
        DefaultAutomaton.Builder.BuilderState hexEscape = builder.newNonFinalState();
        DefaultAutomaton.Builder.BuilderState octalEscape2 = builder.newNonFinalState();
        DefaultAutomaton.Builder.BuilderState gotChar = builder.newNonFinalState();
        DefaultAutomaton.Builder.BuilderState universalEscape = builder.newNonFinalState();
        DefaultAutomaton.Builder.BuilderState gotHexQuad;
        DefaultAutomaton.Builder.BuilderState gotHexQuad2;
        DefaultAutomaton.Builder.BuilderState done = builder.newFinalState();

        init.when(CharConstraint.Builder.eq('\'')).goTo(gotFirstQuote);
        gotFirstQuote.when(CharConstraint.Builder.eq('\\')).goTo(escaped);
        gotFirstQuote.when(CharConstraint.Builder.and(CharConstraint.Builder.inRange(0x20, 128), CharConstraint.Builder.not(CharConstraint.Builder.eq('\\')))).goTo(gotChar);
        escaped.when(CharConstraint.Builder.inList("\"?abfnrtv\\")).goTo(gotChar);

        escaped.when(CharConstraint.Builder.inList(Lexemes.OctalDigit)).goTo(octalEscape);
        octalEscape.when(CharConstraint.Builder.inList(Lexemes.OctalDigit)).goTo(octalEscape2);
        octalEscape.when(CharConstraint.Builder.eq('\'')).goTo(done);
        octalEscape2.when(CharConstraint.Builder.inList(Lexemes.OctalDigit)).goTo(gotChar);
        octalEscape2.when(CharConstraint.Builder.eq('\'')).goTo(done);

        escaped.when(CharConstraint.Builder.eq('x')).goTo(hexEscape);
        escaped.when(CharConstraint.Builder.inList(Lexemes.HexDigit)).goTo(hexEscape);
        hexEscape.when(CharConstraint.Builder.eq('\'')).goTo(done);

        escaped.when(CharConstraint.Builder.or(CharConstraint.Builder.eq('u'), CharConstraint.Builder.eq('U'))).goTo(universalEscape);
        gotHexQuad = addHexQuad(builder, universalEscape);
        gotHexQuad.when(CharConstraint.Builder.eq('\'')).goTo(done);
        gotHexQuad2 = addHexQuad(builder, gotHexQuad);
        gotHexQuad2.when(CharConstraint.Builder.eq('\'')).goTo(done);

        gotChar.when(CharConstraint.Builder.eq('\'')).goTo(done);
        setAutomaton(builder.build());
    }

    private static DefaultAutomaton.Builder.BuilderState addHexQuad(DefaultAutomaton.Builder builder, DefaultAutomaton.Builder.BuilderState origin) {
        DefaultAutomaton.Builder.BuilderState current = origin;
        for (int i = 0; i < 4; i++) {
            DefaultAutomaton.Builder.BuilderState from = current;
            current = builder.newNonFinalState();
            from.when(CharConstraint.Builder.inList(Lexemes.HexDigit)).goTo(current);
        }
        return current;
    }

    public static CCharacter unMarshall(java.io.DataInput in) throws IOException {
        return TerminalImpl.unMarshall(new CCharacter(), in);
    }

    @Override
    public String toString() {
        return "cCharacter";
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (!obj.getClass().equals(CCharacter.class)) {
            return false;
        }
        return true;
    }

    @Override
    public int hashCode() {
        return -14814475;
    }

}

File: src/lexer/CCharacter.ts
import { Terminal } from "../common/Terminal";
import { Automaton, AutomatonBuilder, State } from "./automaton/Automaton";
import { CharConstraint } from "./CharConstraint";

let addHexQuad = (builder:AutomatonBuilder, origin:State):State => {
	let current:State = origin;
	for(let i=0; i<4; i++) {
		let from:State = current;
		current = builder.newNonFinalState();
		builder.addTransition(from, CharConstraint.inList('0123456789abcdefABCDEF'), current);
	}
	return current;
};

let makeAutomaton = (tokenType:Terminal) => {
	let builder = AutomatonBuilder.forTokenType(tokenType);
	let init:State = builder.initialState();
	let gotFirstQuote:State = builder.newNonFinalState();
	let escaped:State = builder.newNonFinalState();
	let octalEscape:State = builder.newNonFinalState();
	let hexEscape:State = builder.newNonFinalState();
	let octalEscape2:State = builder.newNonFinalState();
	let gotChar:State = builder.newNonFinalState();
	let universalEscape:State = builder.newNonFinalState();
	let gotHexQuad:State;
	let gotHexQuad2:State;
	let done:State = builder.newFinalState();

	builder.addTransition(init, CharConstraint.eq('\''), gotFirstQuote);
	builder.addTransition(gotFirstQuote,CharConstraint.eq('\\'), escaped);

	builder.addTransition(
			gotFirstQuote,
			CharConstraint.and(
					CharConstraint.inRange(' ', String.fromCharCode(128)), // 'displayable ascii chars, ie between 0x20 and 0xEF
					CharConstraint.not(CharConstraint.eq('\\'))
			),
			gotChar);

	builder.addTransition(escaped,CharConstraint.inList("\"?abfnrtv\\"), gotChar);

	builder.addTransition(escaped, CharConstraint.inList('01234567'), octalEscape);
	builder.addTransition(octalEscape, CharConstraint.inList('01234567'), octalEscape2);
	builder.addTransition(octalEscape, CharConstraint.eq('\''), done);
	builder.addTransition(octalEscape2, CharConstraint.inList('01234567'), gotChar);
	builder.addTransition(octalEscape2, CharConstraint.eq('\''), done);

	builder.addTransition(escaped, CharConstraint.eq('x'), hexEscape);
	builder.addTransition(escaped, CharConstraint.inList('0123456789abcdefABCDEF'), hexEscape);
	builder.addTransition(hexEscape, CharConstraint.eq('\''), done);

	builder.addTransition(escaped, CharConstraint.inList('uU'), universalEscape);
	gotHexQuad = addHexQuad(builder, universalEscape);
	builder.addTransition(gotHexQuad, CharConstraint.eq('\''), done);
	gotHexQuad2 = addHexQuad(builder, gotHexQuad);
	builder.addTransition(gotHexQuad2, CharConstraint.eq('\''), done);

	builder.addTransition(gotChar, CharConstraint.eq('\''), done);
	return builder.build();
};

class _CCharacter extends Terminal {

	private _automaton:Automaton;

	constructor() {
		super("CCharacter");
		this._automaton = makeAutomaton(this);		
	}

	get automaton() {
		return this._automaton;
	}
};

export const CCharacter:Terminal = new _CCharacter();
